modules/openxr/openxr_api.cpp

```diff
@@ -983,71 +979,84 @@ bool OpenXRAPI::create_swapchains() {
-			depth_views[i].farZ = 100.0;
+	for (uint32_t i = 0; i < render_state.view_count; i++) {
+		render_state.views[i].type = XR_TYPE_VIEW;
+		render_state.views[i].next = nullptr;
+
+		render_state.projection_views[i].type = XR_TYPE_COMPOSITION_LAYER_PROJECTION_VIEW;
+		render_state.projection_views[i].next = nullptr;
+		render_state.projection_views[i].subImage.swapchain = render_state.swapchains[OPENXR_SWAPCHAIN_COLOR].swapchain;
+		render_state.projection_views[i].subImage.imageArrayIndex = i;
+		render_state.projection_views[i].subImage.imageRect.offset.x = 0;
+		render_state.projection_views[i].subImage.imageRect.offset.y = 0;
+		render_state.projection_views[i].subImage.imageRect.extent.width = recommended_size.width;
+		render_state.projection_views[i].subImage.imageRect.extent.height = recommended_size.height;
+
+		if (render_state.submit_depth_buffer && OpenXRCompositionLayerDepthExtension::get_singleton()->is_available() && render_state.depth_views) {
+			render_state.projection_views[i].next = &render_state.depth_views[i];
+
+			render_state.depth_views[i].type = XR_TYPE_COMPOSITION_LAYER_DEPTH_INFO_KHR;
+			render_state.depth_views[i].next = nullptr;
+			render_state.depth_views[i].subImage.swapchain = render_state.swapchains[OPENXR_SWAPCHAIN_DEPTH].swapchain;
+			render_state.depth_views[i].subImage.imageArrayIndex = i;
+			render_state.depth_views[i].subImage.imageRect.offset.x = 0;
+			render_state.depth_views[i].subImage.imageRect.offset.y = 0;
+			render_state.depth_views[i].subImage.imageRect.extent.width = recommended_size.width;
+			render_state.depth_views[i].subImage.imageRect.extent.height = recommended_size.height;
+			render_state.depth_views[i].minDepth = 0.0;
+			render_state.depth_views[i].maxDepth = 1.0;
+			render_state.depth_views[i].nearZ = 0.01; // Near and far Z will be set to the correct values in fill_projection_matrix
+			render_state.depth_views[i].farZ = 100.0;
 		}
 	};
 
 	return true;
 };
 
 void OpenXRAPI::free_swapchains() {
+	// Note, this is called from the render thread when swapchains change
+	// and from the main thread (from destroy_session) when we're cleaning up.
+
 	for (int i = 0; i < OPENXR_SWAPCHAIN_MAX; i++) {
-		if (swapchains[i].swapchain != XR_NULL_HANDLE) {
-			xrDestroySwapchain(swapchains[i].swapchain);
-			swapchains[i].swapchain = XR_NULL_HANDLE;
+		if (render_state.swapchains[i].swapchain != XR_NULL_HANDLE) {
+			if (graphics_extension) {
+				graphics_extension->cleanup_swapchain_graphics_data(&render_state.swapchains[i].swapchain_graphics_data);
+			}
+
+			xrDestroySwapchain(render_state.swapchains[i].swapchain);
+			render_state.swapchains[i].swapchain = XR_NULL_HANDLE;
 		}
 	}
 }
 
 void OpenXRAPI::destroy_session() {
-	if (running && session != XR_NULL_HANDLE) {
-		xrEndSession(session);
-	}
+	// TODO need to figure out if we're still rendering our current frame
+	// in a separate rendering thread and if so,
```

Comment:
@reduz so this is something you might be able to help with, it is possible for a user to shut down OpenXR while we're rendering, which means that we have to make sure we finish rendering the current frame before we continue destroying our sessions, or we will be destroying objects that are still in use.

Is there an easy call into the renderer to pause the main thread until rendering the previous frame has completed?

Reply:
(for the 99% use case the OpenXR session won't be destroyed until we exit Godot, at which time rendering should already be shut down).